[docs] Convert migration guide to asciidoc (#82600)
elastic/kibana: docs/developer/architecture/core/core-api.asciidoc

[[kibana-platform-core-api]]

> == Kibana Platform Core API

Reviewer: ```suggestion
== {kib} Core API
```

experimental[]

> Kibana Platform provides a set of low-level API required to run all the Kibana plugins.
> API is provided as a part of plugin lifecycle contract and may be invoked during an appropriate lifecycle only:

Reviewer: ```suggestion
{kib} Core provides a set of low-level API's required to run all {kib} plugins.
These API's are injected into your plugin's lifecycle methods and may be invoked during that lifecycle only:
```

[source,typescript]
----
import type { PluginInitializerContext, CoreSetup, CoreStart } from 'kibana/server';

export class MyPlugin {
  constructor(initializerContext: PluginInitializerContext) {}

  public setup(core: CoreSetup) {
    // called when plugin is setting up during Kibana's startup sequence
  }

  public start(core: CoreStart) {
    // called after all plugins are set up
  }

  public stop() {
    // called when plugin is torn down during Kibana's shutdown sequence
  }
}
----

== Core services
=== Server-side
[[configuration-service]]
==== Configuration service
Kibana provides `ConfigService` if a plugin developer may want to support
adjustable runtime behavior for their plugins.

> Access to Kibana config in Kibana Platform has been subject to significant refactoring.
> Config service does not provide access to the whole config anymore.
> Kibana platform plugin cannot read configuration parameters of the core services nor other plugins directly.

Reviewer: ```suggestion
Plugins can only read their own configuration values, it is not possible to access the configuration values from {kib} Core or other plugins directly.
```

type ConfigType = TypeOf<typeof configSchema>;

export const config: PluginConfigDescriptor<ConfigType> = {
  exposeToBrowser: {
    uiProp: true,
  },
  schema: configSchema,
};
----

Configuration containing only the exposed properties will be then
available on the client-side using the plugin's `initializerContext`:

[source,typescript]
----
// my_plugin/public/index.ts
interface ClientConfigType {
  uiProp: string;
}

export class MyPlugin implements Plugin<PluginSetup, PluginStart> {
  constructor(private readonly initializerContext: PluginInitializerContext) {}

  public async setup(core: CoreSetup, deps: {}) {
    const config = this.initializerContext.config.get<ClientConfigType>();
  }
----

All plugins are considered enabled by default. If you want to disable
your plugin, you could declare the `enabled` flag in the plugin
config. This is a special Kibana platform key. The platform reads its
value and won’t create a plugin instance if `enabled: false`.

[source,js]
----
export const config = {
  schema: schema.object({ enabled: schema.boolean({ defaultValue: false }) }),
};
----
[[handle-plugin-configuration-deprecations]]
===== Handle plugin configuration deprecations

> If your plugin has deprecated properties, you can describe them using
> the `deprecations` config descriptor field.
> Deprecations are managed on a per-plugin basis, meaning you don’t need to specify
> the whole property path, but use the relative path from your plugin’s
> configuration root.

Reviewer: ```suggestion
If your plugin has deprecated configuration keys, you can describe them using
the `deprecations` config descriptor field.
Deprecations are managed on a per-plugin basis, meaning you don’t need to specify
the whole property path, but use the relative path from your plugin’s
configuration root.
```

      euiIconType: '/path/to/some.svg',
      order: 100,
      appRoute: '/app/my_plugin', // <2>
      async mount(params: AppMountParameters) { // <3>
        // Load application bundle
        const { renderApp } = await import('./application');
        // Get start services
        const [coreStart, depsStart] = await core.getStartServices(); // <4>
        // Render the application
        return renderApp(coreStart, depsStart, params); // <5>
      },
    });
  }
}
----
<1> See {kib-repo}blob/{branch}/docs/development/core/public/kibana-plugin-core-public.applicationsetup.register.md[application.register interface]
<2> Application specific URL.
<3> `mount` callback is invoked when a user navigates to the application-specific URL.
<4> `core.getStartServices` method provides API available during `start` lifecycle.
<5> `mount` method must return a function that will be called to unmount the application.

Note that you are free to use any UI library to render a plugin application in DOM.
However, we recommend using React and https://elastic.github.io/eui[EUI] for all your basic UI
components to create a consistent UI experience.

==== HTTP service

> Provides API to communicate with the Kibana server. Feel free to use another HTTP client library to request 3rd party services.

Reviewer: ```suggestion
Provides API to communicate with the {kib} server. Feel free to use another HTTP client library to request 3rd party services.
```
